pytorch_lightning/trainer/trainer.py

```diff
@@ -796,11 +796,14 @@ def fit(
         if self.use_ddp2:
             task = int(os.environ['SLURM_LOCALID'])
             self.ddp_train(task, model)
-
         elif self.use_ddp:
             if self.is_slurm_managing_tasks:
                 task = int(os.environ['SLURM_LOCALID'])
                 self.ddp_train(task, model)
+            # torchelastic
+            elif 'WORLD_SIZE' in os.environ and 'GROUP_RANK' in os.environ:
```

Comment:
I wonder what we need to do to get this working on slurm. It could be as simple as using the LOCAL_RANK environment instead of the SLURM_LOCALID.

I'll look into it.

Reply:
Okay so looked into it and it's not that simple. Basically, much like distributed training, there's a few ways to initialize elastic training. However, because elastic training needs to own the processes to work, slurm can't spawn them for it.

In distributed training you have the options:
1. run ```python -m torch.distributed.launch <train_script.py>```, which creates the processes for you
2. start the processes yourself using mp (e.g. in the else of this if statement)
3. let slurm (or another scheduler) create the processes

In elastic training you have the options:
1. run ```python3 -m torchelastic.distributed.launch```,  which creates the processes and handles the fault-tolerence and elastic workers.
2. create an elastic agent such as [LocalElasticAgent](https://pytorch.org/elastic/0.2.0/agent.html#torchelastic.agent.server.local_elastic_agent.LocalElasticAgent). This will spawn the elastic processes and manage them with the synchronous function LocalElasticAgent.run().

This doesn't leave a particularly easy way to do slurm because the agent needs to spawn the processes. My guess is that you need to configure slurm to have 1 process per node (i.e. ntasks-per-node=1) and then create the agent and processes as explained in 2. at the beginning of training. You'd also need to setup the distributed key-value store backend (Etcd or Zeus). Luckily they've provided a helpful [python API for spawning Etcd server](https://pytorch.org/elastic/0.2.0/rendezvous.html#torchelastic.rendezvous.etcd_server.EtcdServer).

Reply:
@tullie I don't know what you mean. Lightning already works correctly under a Slurm managed task environment. Do you mean having the same code for both pytorch elastic and slurm?

Reply:
lightning already spawns workers when it detects it is not being managed by slurm

https://github.com/PyTorchLightning/pytorch-lightning/blob/master/pytorch_lightning/trainer/trainer.py#L801-L813

@tullie is this what you mean?

Reply:
![image](https://user-images.githubusercontent.com/3640001/81837628-13f59400-9513-11ea-80c5-4fbd884b68a5.png)

Reply:
yeah, this should work fine no?
If this is the route we take with elastic it means that something else created the process that called each script. Is that the expected behavior @ashwinb @tullie (I haven't used elastic yet).

Reply:
Yes, elastic launches agents on each node which manage the individual worker processes. Lightning's job in that case is to init its process group and configure and run a single trainer worker. This is just like Slurm.

Reply:
All I was saying is that this PR doesn't add support for Elastic Pytorch in a Slurm managed environment. This is fine for now but ideally they'd be able to work together in the future.